helix-core/src/main/java/org/apache/helix/model/InstanceConfig.java

```diff
@@ -331,10 +344,17 @@ public long getInstanceEnabledTime() {
   }
 
   public void setInstanceOperation(InstanceConstants.InstanceOperation operation) {
-    // TODO: also setInstanceEnabled after sanity check.
+    if (operation != InstanceConstants.InstanceOperation.DISABLE
+        && operation != InstanceConstants.InstanceOperation.ENABLE) {
+      if (!getInstanceEnabled()) {
+        throw new HelixException(
+            "setting non enable/disable operation (e.g. evacuate, swap) to helix disabled instance is not allowed");
+      }
+    } else {
+      setInstanceEnabledHelper(operation == InstanceConstants.InstanceOperation.ENABLE);
+    }
```

Comment:
can be simplified as:

if (operation != InstanceConstants.InstanceOperation.DISABLE
         && operation != InstanceConstants.InstanceOperation.ENABLE && !getInstanceEnabled()) {
         throw new HelixException(
            XXX);
     }

setInstanceEnabledHelper(operation == InstanceConstants.InstanceOperation.ENABLE);